dialects/direct/passes/ds2cps.cpp

```diff
@@ -0,0 +1,266 @@
+#include "dialects/direct/passes/ds2cps.h"
+
+#include <iostream>
+
+#include <thorin/lam.h>
+
+#include "dialects/direct/direct.h"
+
+// #define verbose_rewrite
+
+namespace thorin::direct {
+
+/// called by the pass manager when the pass is run
+/// for every lambda in the code
+/// but we need some manual invokation for new lambdas that need urgent care
+void DS2CPS::enter() {
+    Lam* lam = curr_nom();
+    rewrite_lam(lam);
+}
+
+bool isDS(Lam* lam) {
+    auto ty       = lam->type()->as<Pi>();
+    auto codom_ty = ty->codom();
+    return !(codom_ty->isa<Bot>());
+}
+
```

Comment:
There is already a getter `Pi::is_cn`. Also, note that `Lam::type` overrides `Def::type` and already returns `Pi`. So you can remove this one here and simply use it like this:
```c++
isDS(lam)
```
->
```c++
!lam->type()->is_cn()
```